### PR TITLE
making several metrics classes serializable

### PR DESCRIPTION
**Pull request by lbergelson**
making various classes associated with `MetricsFile` serializable

### CODE REVIEW

File: src/java/htsjdk/samtools/util/TestUtil.java [hunk 1]
```diff
@@ -66,4 +65,28 @@ public static void recursiveDelete(final File directory) {
...
+
+        out.writeObject(input);
+        final ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(byteArrayStream.toByteArray()));
+
+        @SuppressWarnings("unchecked")
+        final T result =  (T) in.readObject();
+
+        out.close();
+        in.close();
+        return  result;
```
Comment (nh13) [inline src/java/htsjdk/samtools/util/TestUtil.java]: two spaces </nitpick>

File: src/tests/java/htsjdk/samtools/metrics/StringHeaderTest.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+public class StringHeaderTest {
+
+    @Test
+    public void testStringHeaderSerialization() throws IOException, ClassNotFoundException {
+        Header header = new StringHeader("some value");
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/metrics/StringHeaderTest.java]: ```final``` here and below?

File: src/tests/java/htsjdk/samtools/metrics/VersionHeaderTest.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+
+public class VersionHeaderTest {
+
+    @Test
+    public void testSerializeVersionHeader() throws IOException, ClassNotFoundException {
+        VersionHeader versionHeader = new VersionHeader();
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/metrics/VersionHeaderTest.java]: ```final```

File: src/tests/java/htsjdk/samtools/util/HistogramTest.java [hunk 4]
```diff
@@ -76,6 +78,18 @@ public void testMad() {
     }
 
 
+    @Test(dataProvider = "histogramData")
+    public void testSerializeHistogram(final int[] values, final double unused_mean, final double unused_stdev, final Integer unused_trimByWidth) throws IOException, ClassNotFoundException {
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/util/HistogramTest.java]: what are these variables with underscores?

File: src/tests/java/htsjdk/samtools/util/HistogramTest.java [hunk 5]
```diff
@@ -76,6 +78,18 @@ public void testMad() {
...
+        final Histogram<Integer> histo = new Histogram<Integer>();
+        for (int value : values) {
+            histo.increment(value);
+        }
+
+        Histogram<Integer> deserializedHistogram = TestUtil.serializeAndDeserialize(histo);
+        Assert.assertEquals(deserializedHistogram, histo);
+
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/util/HistogramTest.java]: every blank line is precious

**Comment by nh13** [comment]
minor stuff, but :+1:

File: src/java/htsjdk/samtools/util/TestUtil.java (hunk 1 above)
Comment (lbergelson) [inline src/java/htsjdk/samtools/util/TestUtil.java]: how do you even notice these?  Eagle-eyes...

File: src/tests/java/htsjdk/samtools/util/HistogramTest.java (hunk 4 above)
Comment (lbergelson) [inline src/tests/java/htsjdk/samtools/util/HistogramTest.java]: Yeah, those are weird... They're provided by the data provider, but we don't use them for anything in this tests, so I gave them obvious strange names to let people know.

I'll add a comment stating they're unused and change them to just be `mean`.

**Comment by lbergelson** [comment]
@nh13 Thanks for the quick review!

File: src/java/htsjdk/samtools/util/TestUtil.java [hunk 6]
```diff
@@ -66,4 +65,28 @@ public static void recursiveDelete(final File directory) {
...
+     */
+    public static <T extends Serializable> T serializeAndDeserialize(T input) throws IOException, ClassNotFoundException {
+        final ByteArrayOutputStream byteArrayStream = new ByteArrayOutputStream();
+        final ObjectOutputStream out = new ObjectOutputStream(byteArrayStream);
+
+        out.writeObject(input);
+        final ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(byteArrayStream.toByteArray()));
+
+        @SuppressWarnings("unchecked")
+        final T result =  (T) in.readObject();
```
Comment (nh13) [inline src/java/htsjdk/samtools/util/TestUtil.java]: eagle eyes

Comment (lbergelson) [inline src/java/htsjdk/samtools/util/TestUtil.java]: if you have a preferred linter I can just run that as a pre-commit hook.

**Pull request merged**